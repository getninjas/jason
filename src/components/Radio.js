import PropTypes from 'prop-types';
import React, { Component } from 'react';

const propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onFieldChange: PropTypes.func.isRequired,
  onFieldBlur: PropTypes.func.isRequired,
  type: PropTypes.string,
  required: PropTypes.bool,
  values: PropTypes.array,
};

const defaultProps = {
  required: false,
  type: 'radio',
  values: [],
};

// TODO: test
export default class Radio extends Component {
  constructor(props) {
    super(props);

    this.state = {}

    this.onChange = this.onChange.bind(this);
    this.onBlur = this.onBlur.bind(this);
  }

  // TODO: test
  onChange(evt) {
    const value = evt.target.value;
    this.setState({ value });

    this.props.onFieldChange({
      value,
      id: this.props.id
    });
  }

  onBlur() {
    this.props.onFieldBlur({
      value: this.state.value,
      id: this.props.id,
      required: this.props.required,
    });
  }

  render() {
    const {
      name,
      required,
      type,
    } = this.props;

    return (
      <ul>
        {this.props.values.map((elem, idx) => (
          <li className='form__radio' key={`${elem.databaseId}-${idx}`}>
            <input
              type={type}
              id={elem.databaseId}
              name={name}
              value={elem.databaseId}
              onChange={this.onChange}
              onBlur={this.onBlur}
              required={required} />
            <label htmlFor={elem.databaseId}>{elem.value}</label>
          </li>
        ))}
      </ul>
    );
  }
}

Radio.propTypes = propTypes;
Radio.defaultProps = defaultProps;
