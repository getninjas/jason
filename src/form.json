{
  "form": {
    "title": "Form Title",
    "databaseId": 81,
    "zipcodeUrlService": "http://www.mocky.io/v2/5afaebc92f00000e00f7c654",
    "steps": [
      {
        "title": "General data",
        "breadcrumbTitle": "Your request",
        "buttonText": "Start request",
        "headerMarkup": "<h2 class='widget__title section-title--high-contrast'>Lorem Ipsum is simply dummy text <small>It has survived not</small></h2>",
        "fields": [
          {
            "title": "Qual é o tipo de serviço que você precisa?",
            "placeholder": "Placeholder",
            "required": true,
            "id": "1_id",
            "name": "1_name_select",
            "type": "select",
            "value": 7117,
            "values": [
              {
                "databaseId": 7117,
                "value": "Construção"
              },
              {
                "databaseId": 7118,
                "value": "Reformas"
              },
              {
                "databaseId": 7119,
                "value": "Instalações"
              },
              {
                "databaseId": 7120,
                "value": "Troca"
              },
              {
                "databaseId": 7121,
                "value": "Outros"
              }
            ]
          },
          {
            "title": "Especifique melhor a metragem do serviço",
            "required": true,
            "id": "2_id",
            "name": "2_name_text",
            "type": "text",
            "placeholder": "Bora pra action",
            "value": "AHUAHuhhHUHAU",
            "values": []
          },
          {
            "title": "Informações adicionais",
            "required": false,
            "type": "textarea",
            "id": "3_id",
            "name": "3_name_textarea",
            "placeholder": "Descreva o que você precisa",
            "value": null,
            "values": []
          },
          {
            "required": true,
            "title": "Você precisa de serviços adicionais?",
            "placeholder": "Você precisa de serviços adicionais?",
            "id": "0_id",
            "name": "1_name_select",
            "type": "checkbox",
            "value": null,
            "values": [
              {
                "databaseId": 6792,
                "value": "Cozinhar"
              },
              {
                "databaseId": 7639,
                "value": "OTHER"
              }
            ]
          }
        ]
      },
      {
        "title": "User Data",
        "breadcrumbTitle": "Your data",
        "buttonText": "Finish",
        "headerMarkup": "<div class='form__alert-message space-element-large'><img src='https://www.getninjas.com.br/assets/request/phone_call.svg' class='form__icon' alt='Lorem Ipsum is simply dummy text of the printing and typesetting industry.'><h2 class='form__info'>Almost there<small>Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's<strong>standard dummy text ever since the 1500s</strong></small></h2></div>",
        "fields": [
          {
            "title": "Zipcode",
            "required": true,
            "type": "zipcode",
            "id": "zip_code",
            "name": "zipcode",
            "placeholder": "00000-000",
            "value": "04904-060",
            "wrapperClassName": "form__field input",
            "values": []
          },
          {
            "title": "DDD + Celular",
            "id": "phone",
            "name": "phone",
            "type": "phone",
            "placeholder": "(__) _____-____",
            "required": true,
            "value": "(11) 96311-2689",
            "values": []
          }
        ]
      }
    ]
  }
}
