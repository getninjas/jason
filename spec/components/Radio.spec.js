import React from 'react';
import renderer from 'react-test-renderer';
import { form } from '../../src/form.json';
import Radio from '../../src/components/Radio';
import { enzymeConfig, mount } from '../enzymeConfig';

enzymeConfig();

function createNodeMock(element) {
  if (element.type === 'radio') {
    return {
      addEventListener() {},
      values: '',
    };
  }

  return null;
}

const radio = form.steps[0].fields[0];

describe('Radio', () => {
  const commonProps = {
    id: 'idTest',
    name: 'nameTest',
    onFieldChange: () => {},
    onFieldBlur: () => {},
    placeholder: 'placeholderTest',
    type: 'radio',
    required: true,
  };

  it('renders defaultProps', () => {
    const options = { createNodeMock };
    const component = renderer.create(
      <Radio
        {...commonProps}
        values={[radio.values[0]]}
      />, options,
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('.onChange', () => {
    const onChange = jest.fn();

    const component = mount(
      <Radio
        {...commonProps}
        values={[radio.values[0]]}
        onFieldChange={onChange}
      />,
    );

    component.find('input').simulate('change', { target: { value: 6792, id: 'idTest' } });

    expect(component.instance().props.onFieldChange).toBeCalledWith({ value: 6792, id: 'idTest' });
  });

  it('.onBlur', () => {
    const onBlur = jest.fn();

    const component = mount(
      <Radio
        {...commonProps}
        values={[radio.values[0]]}
        onFieldBlur={onBlur}
      />,
    );

    component.find('input').simulate('blur');

    expect(component.instance().props.onFieldBlur).toBeCalled();
  });
});
