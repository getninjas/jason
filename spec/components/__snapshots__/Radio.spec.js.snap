// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio renders defaultProps 1`] = `<ul />`;
